### ISSUE TITLE
ON HOLD: Export/import as/from spreadsheet (XLSX, ODS) file

### ISSUE THREAD

**Issue by georgeef** [moneymanagerex/moneymanagerex#6837]
This is a proposal for a new interchange format in XLSX. I have implemented this feature in https://github.com/georgeef/mmexutil (written in Perl). I can start working on a native implementation in MMEX if there is enough interest. I will need some support, mostly at DB level, later maybe also at gui level.

## Motivation

The interchange formats currently supported by MMEX, like CSV and QIF, are quite abstract and simple, however they do now cover many of the details in MMEX database. In particular, exporting to any of these formats and then importing again to an empty MMEX database, can hardly reproduce the original database. I tried all of them, in order to transfer my data to MMEX, but I was not satisfied by the result. There are regular problem reports and requests for enhancement in export/import functionality from users.

The MMEX (.mmb) format itself is an sqlite database, which is an open and standard format, so in principle olny an sqlite driver is needed in order to read from and write into it. However, most users cannot handle sqlite files directly, and the effort to implement all or most of the MMEX details (e.g., database invariants) is not trivial.

## Features

The proposed XLSX format is a much more detailed representation (essentially a lossless dump) of the MMEX database. Only the rowids are ignored (as it should happen in any interchange format), but the relationships between records, defined through rowids, are preserved.

Most users can easily handle XLSX files, including transfer of data from/to other sources. Of course they need to comply with the syntax and keywords of this new format, but this is easier than, or as easy as, with other formats.

Strictly speaking, the proposed XLSX format is more of a database dump than an abstract interchange format, since it mirrors the MMEX database schema. From my experience, with a little effort (editing data in Excel/LbreOffice) it can serve the purposes of interchange format quite well.

## Use cases

I implemented this feature after I used MMEX for some time with new (real) data, and then decided to switch to it, therefore I had to merge the new data in MMEX with old historical data in another format. Although this transfer was executed only once, it was a major step in order to be able to switch to MMEX. I guess that most people who used other tools in the past, would like to execute this kind of transfer in a complete and reliable way, at least to not lose data that are compatible with or convertible to MMEX.

At the time I implemented the proposed format, I needed only the import functionality. The export functionality was developed as an intermediate step, and in order to be used as a check of the transfer. Now that I have most of my data in MMEX, I feel more confortable that at any time I can export to a convenient format, although I don't need this functionality.

The proposed format can also be used as an intermediate step in order to merge two MMEX databases, e.g., the newer database can be exported to XLSX and then imported into the older database.

## Future of my implementation

I hope that the proposed format, or something equivalent, can be integrated into MMEX, so (1) it can be useful for others, and (2) I won't need to maintain my own tool in Perl, which requires tracking of the progress in MMEX.

At the moment I use my tool for 1-2 minor features (mostly cash flow report), which I cannot find in MMEX. I'll propose them in another issue.

**Comment by georgeef** [comment]
## Task list

- [ ] Decide which external library to use. Check license compatibility.
- [ ] Test external library on macOS.
- [ ] Add basic export/import functionality in Debug mode (or with feature flag).
- [ ] Test external library with basic functionality on Linux and on Windows (support needed).
- [ ] Implement complete export functionality.
- [ ] Develop dialog form for the export options.
- [ ] Enable export functionality in Release mode.
- [ ] Update `BUILD.md` for macOS.
- [ ] Update `BUILD.md` for Linux and for Windows (support needed).
- [ ] Implement import functionality, in small parts for easier review (bugs in import may corrupt the database).
- [ ] Analyze possible errors during import. Protect database against import errors.
- [ ] Develop dialog form for the import options.
- [ ] Enable import functionality in Release mode.
- [ ] Refactor and cleanup.

**Comment by georgeef** [comment]
Yes, the XLSX file needs to be formatted in a specific way.

You can try `mmexutil` to see the format, e.g., create an empty database with MMEX and then export it to XLSX. This will show the default worksheet names and the (fixed) column names. It will not show the keyword values (several columns accept only special keywords as values). You can browse through the source code to see all details, but in practice if you export an already populated database you see most of the keywords, or at least those that you need.

Of course the format may be adapted for the implementation in MMEX, but the idea is the same.

**Comment by jensb** [comment]
Interesting.
I wrote a converter for the iFinance (MacOS) application database (also SQLite but different schema) to MMEX for my own purposes, I can publish it if there is interest. I am also planning to write a data importer that can batch-query various APIs (FinTS using aqbanking, Paypal, ...) and collect data from there.
Maybe we should document the various MMEX utilities somewhere?

**Comment by georgeef** [comment]
I am not aware of other MMEX utilities, but I agree, if there are a few, it makes sense to collect them in a document (maybe a section in README). Such utilities tend to become outdated though, unless they track the development in the main application.

**Comment by ovari** [comment]
Should `to` be `as`?

Should the title be changed from
`Export/import to/from XLSX file`
to
`Import from XLSX File…`/`Export as XLSX File…`
or
`Export/import as/from XLSX File…`

Thank you

**Comment by ovari** [comment]
Can the default [LibreOffice](https://www.libreoffice.org/) [Calc spreadsheet](https://www.libreoffice.org/discover/calc/) [Open Document Format (ODF)](https://en.wikipedia.org/wiki/OpenDocument) format please be supported, `ODF spreadsheet (.ods)`, please be supported?
`Import from ODS File…`/`Export as ODS File…`

Thank you

**Comment by ovari** [comment]
> ## Task list
>     [ ]  Decide which external library to use. Check license compatibility.

Does [SheetJS](https://sheetjs.com/) help?

Thank you

**Comment by georgeef** [comment]
ODS can be supported if the library supports it, without much code re-write. it could be the default or the only supported format. i picked XLSX because it might be more convenient for Excel users.

the features needed for this export/import are:
- minimum: text, numbers, font selection (dates can be processed as text).
- good to have: header (rows) freeze.
- good to have: group/outline (in both columns and rows).

i tested with LibreOffice, it supports all these features with XLSX very well.

**Comment by georgeef** [comment]
> Does [SheetJS](https://sheetjs.com/) help?

no, this requires a JS engine. i need a plain library in C++, or (not preferable) in C.

[libxlsxwriter](https://github.com/jmcnamara/libxlsxwriter) is a C library, it does not support read.

[xlnt](https://github.com/xlnt-community/xlnt) is a C++ library, i didn't test it yet.

**Comment by luisdev** [comment]
@georgeef You closed this issue - what happened? Are you not doing this anymore?

**Comment by georgeef** [comment]
See discussion in #6879.
